src/algorithms/pagerank.c

```diff
@@ -0,0 +1,224 @@
+#include <assert.h>
+
+//------------------------------------------------------------------------------
+// scalar operators
+//------------------------------------------------------------------------------
+
+#define DAMPING 0.85
+
+void fdiff(void *z, const void *x, const void *y) {
+	float delta = (* ((float *) x)) - (* ((float *) y)) ;
+	(*((float *) z)) = delta * delta ;
+}
+
+//------------------------------------------------------------------------------
+// comparison function for qsort
+//------------------------------------------------------------------------------
+
+int compar(const void *x, const void *y) {
+	LAGraph_PageRank *a = (LAGraph_PageRank *) x ;
+	LAGraph_PageRank *b = (LAGraph_PageRank *) y ;
+
+	// sort by pagerank in descending order
+	if(a->pagerank > b->pagerank) {
+		return (-1) ;
+	} else if(a->pagerank == b->pagerank) {
+		return (0) ;
+	} else {
+		return (1) ;
+	}
+}
+
+//------------------------------------------------------------------------------
+// LAGraph_pagerank: compute the pagerank of all nodes in a graph
+//------------------------------------------------------------------------------
+
+GrB_Info Pagerank               // GrB_SUCCESS or error condition
+(
+	LAGraph_PageRank **Phandle, // output: array of LAGraph_PageRank structs
+	GrB_Matrix A,               // binary input graph, not modified
+	int itermax,                // max number of iterations
+	double tol,                 // stop when norm (r-rnew,2) < tol
+	int *iters                  // number of iterations taken
+) {
+
+	//--------------------------------------------------------------------------
+	// initializations
+	//--------------------------------------------------------------------------
+
+	GrB_Info info ;
+	float rsum ;
+	float *X = NULL ;
+	LAGraph_PageRank *P = NULL ;
+	GrB_BinaryOp op_diff = NULL ;
+	GrB_Index n, nvals, *I = NULL ;
+	GrB_Vector r = NULL, t = NULL, d = NULL ;
+	GrB_Matrix C = NULL, D = NULL, T = NULL ;
+
+	assert(Phandle);
+	(*Phandle) = NULL ;
+
+	// n = size (A,1) ;         // number of nodes
+	assert(GrB_Matrix_nrows(&n, A) == GrB_SUCCESS) ;
+	if(n == 0) return (GrB_SUCCESS) ;
+
+	// teleport = (1 - 0.85) / n
+	float one = 1.0 ;
+	float teleport = (one - DAMPING) / ((float) n) ;
+
+	// r (i) = 1/n for all nodes i
+	float x = 1.0 / ((float) n) ;
+	assert(GrB_Vector_new(&r, GrB_FP32, n) == GrB_SUCCESS) ;
+	assert(GrB_assign(r, NULL, NULL, x, GrB_ALL, n, NULL) == GrB_SUCCESS) ;
+
+	// d (i) = out deg of node i
+	assert(GrB_Vector_new(&d, GrB_FP32, n) == GrB_SUCCESS) ;
+	assert(GrB_reduce(d, NULL, NULL, GrB_PLUS_FP32, A, NULL) == GrB_SUCCESS) ;
+	// GxB_print (d, 3) ;
+
+	// D = (1/diag (d)) * DAMPING
+	GrB_Type type ;
+	assert(GxB_Vector_export(&d, &type, &n, &nvals, &I, (void **)(&X), NULL) == GrB_SUCCESS) ;
+
+	for(int64_t k = 0 ; k < nvals ; k++) X [k] = DAMPING / X [k] ;
+	assert(GrB_Matrix_new(&D, GrB_FP32, n, n) == GrB_SUCCESS) ;
+	assert(GrB_Matrix_build(D, I, I, X, nvals, GrB_PLUS_FP32) == GrB_SUCCESS) ;
```

Comment:
This assertion fails if the relationship is set to something invalid and is replaced with the zero matrix.

Reply:
👍